feat(rust/rbac-registration): refactoring

stanislav-tkach opened the pull request:
# Description

- The `Cip0134UriSet` type was added.
- `Option<Vec<_>>` fields of the `Cip509RbacMetadata` structure were changed to `Vec<_>`.
- Error handling becomes stricter.
- Store x509 certificates (Cip509RbacMetadata::x509_certs) in the decoded format.
- Store certificate URIs in the registration chain.
- Use types from the `cardano-blockchain-types` crate.

## Related Issue(s)

Closes https://github.com/input-output-hk/catalyst-libs/issues/104.

## Related Pull Requests

This is a combination of the following pull requests:
- https://github.com/input-output-hk/catalyst-libs/pull/119
- https://github.com/input-output-hk/catalyst-libs/pull/128
- https://github.com/input-output-hk/catalyst-libs/pull/130

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

bkioshn commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -92,7 +107,13 @@ pub(crate) enum Cip509IntIdentifier {
         let map_len = decode_map_len(d, "CIP509")?;
-        let mut cip509_metadatum = Cip509::default();
+
+        let mut purpose = Uuid::default();
+        let mut txn_inputs_hash = TxInputHash::default();
```
Maybe better to set "not set" value to `None`.
Or check whether these required fields are set or not, if no return Error.

Can think about it in other PR since it is not related to this PR .

stevenj replied:
Agree, don't assume that the fields are some default if they are not decoded, they should be set to a "not set" state, typically by using `Option<>` and defaulting them to None.

stanislav-tkach replied:
I totally agree, but previously the whole `Cip509` structure was constructed in the default state, so I wasn't sure if it is ok.

bkioshn reviewed (dismissed):
LGTM 🌈

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -25,28 +25,43 @@ use validation::{
+    /// A registration purpose (`UUIDv4`).
+    ///
+    /// The purpose is defined by the consuming dApp.
+    pub purpose: Uuid,
```
I would rather that none of the field are public, and we expose methods which use them where required.

stanislav-tkach replied:
Usually I also prefer this approach, but in this case we move values out of `Cip509` fields when constructing the registration chain. With private fields we will be forced to clone them. Probably not a big dial, but still fills unnecessary.

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -25,28 +25,43 @@ use validation::{
+    ///
+    /// The hash must always be present except for the first registration transaction.
+    // TODO: Use the `Blake2b256Hash` type from the `cardano-blockchain-types` crate.
+    pub prv_tx_id: Option<Hash<32>>,
```
That crate is merged now, so this should use that type.

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -101,43 +122,57 @@ impl Decode<'_, ()> for Cip509 {
-                                decode::Error::message("Invalid data size of Purpose")
-                            })?;
+                        purpose = Uuid::try_from(decode_bytes(d, "CIP509 purpose")?)
+                            .map_err(|_| decode::Error::message("Invalid data size of Purpose"))?;
```
Gather all errors on decode, don't fail fast.

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -25,28 +25,43 @@ use validation::{
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/blob/x509-envelope-metadata/CIP-XXXX/README.md
+#[derive(Debug, PartialEq, Clone)]
 pub struct Cip509 {
```
I would rather we have a `new` method that takes a `MultiEraBlock` and a `TxnIndex` and it returns a Cip509 struct from it.  The only reason it would return a "Error" is if there was no identifiable Cip509 data in the block at that index, otherwise it decodes what it can, and returns a Cip509 struct which is decoded as much as possible, and if it had errors, includes a validation report which is not empty which lists all errors found during deserialization or validation.

Also, I don't believe you can meaningfully decode or validate the RBAC data without the full transaction.

A helper function could then take just a `MultiEraBlock`, iterate it for `TxnIndex` and return a `Vec<(TxnIndex,Cip509)>` for all Cip509 metadata found in the block.  And an empty vec if none is found.

This is almost always how we would deserialize the Cip509 data for indexing, as all indexing is done at the block level.

There are Two primary API's any consumer would want from this type:
1. Construction of a single rbac registration metadata from a transaction or a set of them from a block (as described above).
2. Adding the next sequential decoded Cip509 transaction, to the current one to create a unified C509 transaction from the two, (and validate that they are able to be added in that way).

The only other Api's should be to query the contents of the Cip509 registration as currently known.

All the inner details should be accessed only through methods on this Cip509 structure.

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -179,16 +214,14 @@ impl Cip509 {
-                    is_valid_signing_key = validate_role_singing_key(role, validation_report);
-                }
+        // Validate only role 0
+        for role in &self.metadata.role_set {
```
Should not need to iterate here, should be able to reference it directly by role number.
This is a very common behaviour.  the Cip509 struct should have a method to return the roles it defines, and then to query information from any role it contains.

conceptually:
```
self.metadata.role_set.get(0)
```
should return `Some` data if its in the registration or `None` if its not.
And we should be able to do something like:
```
self.metadata.role_set.all_roles() -> Vec<RoleNumber>
```
To get a list of all Role Values defined in the registration.

As said above, we should use this internally, and it should also be exposed public API for a consumer.

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -16,24 +16,43 @@ use role_data::RoleData;
+    /// A potentially empty list of revoked certificates.
+    pub revocation_list: Vec<CertKeyHash>,
+    /// A potentially empty list of role data.
+    pub role_set: Vec<RoleData>,
```
Whiles its a Vec in the CBOR, there is no need to 100% preserve that format in the structure.  It should be decoded into a map.  The Key should be the Role Number, and the Value the RoleData.  We should not use `u8` as an implcit type for the `RoleNumber` we should make a `RoleNumber` type to contain it.

We should then make a set of constants of type `RoleNumber` which define our known Role Numbers, Role0 being one of those constants, rather than always using `0` as an implicit value.

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -60,86 +79,59 @@ pub enum Cip509RbacMetadataInt {
                     Cip509RbacMetadataInt::X509Certs => {
-                        let x509_certs = decode_array_rbac(d, "x509 certificate")?;
-                        x509_rbac_metadata.set_x509_certs(x509_certs);
+                        x509_certs = decode_array_rbac(d, "x509 certificate")?;
```
Don't fail fast, try and decode, and record an error if its not decodable.

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -60,86 +79,59 @@ pub enum Cip509RbacMetadataInt {
                     Cip509RbacMetadataInt::C509Certs => {
-                        let c509_certs = decode_array_rbac(d, "c509 certificate")?;
-                        x509_rbac_metadata.set_c509_certs(c509_certs);
+                        c509_certs = decode_array_rbac(d, "c509 certificate")?;
```
ditto

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -60,86 +79,59 @@ pub enum Cip509RbacMetadataInt {
                     Cip509RbacMetadataInt::PubKeys => {
-                        let pub_keys = decode_array_rbac(d, "public keys")?;
-                        x509_rbac_metadata.set_pub_keys(pub_keys);
+                        pub_keys = decode_array_rbac(d, "public keys")?;
```
ditto

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -60,86 +79,59 @@ pub enum Cip509RbacMetadataInt {
                     Cip509RbacMetadataInt::RevocationList => {
-                        let revocation_list = decode_revocation_list(d)?;
-                        x509_rbac_metadata.set_revocation_list(revocation_list);
+                        revocation_list = decode_revocation_list(d)?;
```
ditto

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/mod.rs:
```diff
@@ -60,86 +79,59 @@ pub enum Cip509RbacMetadataInt {
                     Cip509RbacMetadataInt::RoleSet => {
-                        let role_set = decode_array_rbac(d, "role set")?;
-                        x509_rbac_metadata.set_role_set(role_set);
+                        role_set = decode_array_rbac(d, "role set")?;
```
ditto

stevenj commented on rust/rbac-registration/src/cardano/cip509/mod.rs:
```diff
@@ -92,7 +107,13 @@ pub(crate) enum Cip509IntIdentifier {
 impl Decode<'_, ()> for Cip509 {
     fn decode(d: &mut Decoder, ctx: &mut ()) -> Result<Self, decode::Error> {
```
The metadata should take the paired transaction and witness set as context.  Its not possible to decode properly without it, i don't think?

bkioshn replied:
It is needed in validation and collecting the data for the registration chain.

stevenj commented on rust/rbac-registration/src/cardano/cip509/utils/cip134/mod.rs:
```diff
@@ -0,0 +1,7 @@
+//! Utilities for [CIP-134] (Cardano URIs - Address Representation).
```
Move the `Cip0134Uri` type to the `cardano-blockchain-types` crate.  Its not exclusively part of RBAC, its just used by RBAC.

stevenj commented on rust/rbac-registration/src/cardano/cip509/utils/cip134/uri_set.rs:
```diff
@@ -0,0 +1,316 @@
+                .content
+                .as_slice()
+                .with_context(|| "Unable to process content for {data:?}")?;
+            let address = match decode_utf8(content) {
+                Ok(a) => a,
+                Err(e) => {
+                    // X.509 doesn't restrict the "alternative name" extension to be utf8 only, so
+                    // we cannot treat this as error.
+                    debug!("Ignoring {e:?}");
+                    continue;
+                },
+            };
+            let uri = match Cip0134Uri::parse(&address) {
+                Ok(u) => u,
+                Err(e) => {
+                    // Same as above - simply skip non-confirming values.
+                    debug!("Ignoring invalid CIP-0134 address: {e:?}");
+                    continue;
+                },
+            };
```
Feel like this should be a method of `Cip509URI` that tries to make it from a `slice` of bytes, rather than done inline here.

stevenj commented on rust/rbac-registration/src/cardano/cip509/validation.rs:
```diff
@@ -273,6 +120,24 @@ pub(crate) fn validate_stake_public_key(
 
+    // TODO: This should be improved, probably along with the `compare_key_hash` function
+    // implementation.
+    let pk_addrs: Vec<_> = cip509
```
This needs a comment explaining exactly what its doing, its not possible to read it and know what its doing (or supposed to be doing).  Also looks kind of complex, so its probably a good candidate to be its own function (that explains what its doing) and a unit test which tests that it does what its aiming to do.

stanislav-tkach commented:
I think that I have addressed all the comments, so I'm marking the pull request as ready to review. There are still a few things that can still be improved, so I will try to do that during one more round of self-review. But I believe that generally the code becomes better especially considering error reporting.

Now I'm mostly concerned about the registration chain structure both internal parts and the interface. For example `RegistrationChain::new` takes transaction id and point and we already have these things in Cip509. @stevenj We discussed that applying `Cip509` to the registration chain should be infallible (except of the previous transaction hash check) if a new registration is valid. I haven't fully implemented that because I don't want to change the `RegistrationChain` structure before confirming that I'm moving in a right direction. I will leave a few comments as a points for discussion.

Also I will look into the coverage report and try to increase it.

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/rbac/metadata.rs:
```diff
@@ -0,0 +1,230 @@
+    /// A potentially empty list of revoked certificates.
+    pub revocation_list: Vec<CertKeyHash>,
+    /// A potentially empty role data.
+    pub role_data: HashMap<RoleNumber, RoleData>,
```
We already store the role data as a map in the registration chain, so I think here it would be better to revert my changes and have a structure that follows CBOR format just to simplify decoding unless we want to have a map here too.

bkioshn commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,497 @@
+};
+use pallas::{
+    codec::utils::Nullable,
+    ledger::traverse::{MultiEraBlock, MultiEraTx},
```
Let's use our `MultiEraBlock` , the current one doesn't have functionality that you want, but will add one

bkioshn commented on rust/rbac-registration/src/utils/decode_helper.rs:
```diff
@@ -1,5 +1,8 @@
 //! CBOR decoding helper functions.
```
This file is currently move out to `cbork-utils`

bkioshn reviewed (commented):
If the decode sequence is not what we expected, we should not be able to process any further.
for example if we expect [1, 2, 3] 
```
83    # array(3)
   01 # unsigned(1)
   02 # unsigned(2)
   03
```
and the data instead be 1
```
01 # unsigned(1)
```
This will corrupt the everything else and hence we cannot process any further

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/rbac/metadata.rs:
```diff
@@ -0,0 +1,230 @@
+                    },
+                    Err(e) => {
+                        report.other(&format!("Unable to decode purpose value: {e:?}"), context);
+                    },
```
It is a decoding problem, but I'm not sure about stopping. Lets discuss the error handling tomorrow. 🙃

bkioshn commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,569 @@
+
+    for (index, output) in txn.transaction_body.outputs.iter().enumerate() {
+        let conway::PseudoTransactionOutput::PostAlonzo(output) = output else {
+            report.other("Unsupported legacy transaction output", &context);
```
Since we are recording the payment address that we care, not sure whether we want to report the problem here, if it somehow legacy outputs, we just don't care about it. WDYT?

stanislav-tkach replied:
I have followed the [previous logic](https://github.com/input-output-hk/catalyst-libs/blob/main/rust/rbac-registration/src/registration/cardano/mod.rs#L599), but I think that you are right. I will change that, but just in case let's ask @stevenj.

bkioshn commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,569 @@
+        let address = match Address::from_bytes(&output.address) {
+            Ok(Address::Shelley(a)) => a,
+            Ok(a) => {
+                report.other(&format!("Unexpected output address type: {a:?}"), &context);
```
This one too

bkioshn commented on rust/rbac-registration/src/utils/general.rs:
```diff
@@ -1,34 +1,23 @@
+
 /// Getting the index by decrementing by 1.
 /// e.g. 1 should refers to index 0
 pub(crate) fn decremented_index(int: i16) -> anyhow::Result<usize> {
```
This can be removed, since payment-key doesn't support both tx inputs and puts

bkioshn commented on rust/rbac-registration/src/utils/hashing.rs:
```diff
@@ -2,14 +2,6 @@
-}
-
 /// Convert the given value to `blake2b_256` array.
 pub(crate) fn blake2b_256(value: &[u8]) -> anyhow::Result<[u8; 32]> {
```
I think the hashing function is available in other cat-lib crate, need to recheck

stevenj replied:
Yes, its generalized into another crate, shouldn't be here anymore.

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,569 @@
+            )
+        })?;
+        let MultiEraTx::Conway(txn) = txn else {
+            return Err(anyhow!("Unsupported era: {}", txn.era()));
```
I agree, this makes sense.

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,569 @@
+        // Find the transaction and decode the relevant data.
+        let txns = block.txs();
+        let txn = txns.get(usize::from(index)).ok_or_else(|| {
+            anyhow!(
```
I'm not sure about that: it is a developer error that can only occur if someone calls `Cip509::new` with an invalid index and it should never happen when `Cip509::from_block` is called. I think we shouldn't ignore such errors.la

bkioshn replied:
Aw okay, misunderstand the code

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,569 @@
+            txn.transaction_body.auxiliary_data_hash.as_ref(),
+            &cip509.report,
+        );
+        // The following checks are only performed for  the role 0.
```
Good catch, I'm still adjusting to a new keyboard. 🙃

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/types/role_data.rs:
```diff
@@ -0,0 +1,221 @@
+    /// A signing key.
+    signing_key: Option<KeyLocalRef>,
+    /// An encryption key.
+    encryption_key: Option<KeyLocalRef>,
```
I think this is probably a good idea, but I have a related question. Both [before](https://github.com/input-output-hk/catalyst-libs/blob/main/rust/rbac-registration/src/cardano/cip509/mod.rs#L185) and [currently](https://github.com/input-output-hk/catalyst-libs/pull/127/files#diff-230d6c465d9430867e64aa10f0fc88365f64fb632a1ea4a5896189d0a7d60ae2R173) we check the signing key only for the role 0. I still don't fully understand that. Should we check and convert them for all roles?.. Is it possible to reference a "missing" key? For example, to have `X509DerCert::Undefined` in a registration and reference a key from this certificate?

Also we don't check the encryption keys at all.

@stevenj Help. 🙃

bkioshn replied:
Role 0 must have signing key and it should be referencing a certificate, but for other role, it is not restricted. For encryption key there is no restriction, since there is no use case for it yet.

bkioshn replied:
And for undefined cert/key, possible, but we need to look at the registration chain to see what is at that index.

stevenj replied:
We will also know role 0,  its possible we will have a ref, but the key is in an earlier transaction, so we won't be able to deref them until we accumulate the registrations in the chain.
However when we do accumulate, it would be an error if there was a ref and it didn't have a key in it.

stevenj replied:
For now, I would be inclined to leave this like this, and just keep in mind as we iterate and implement RBAC stuff in cat-gateway that we might need to improve this in some way, rather than second guess it right now.

stanislav-tkach replied:
I have improved a verification of the role 0 signing key slightly.

stevenj commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,497 @@
+        let mut report = ProblemReport::new("Decoding and validating Cip509");
+        let mut cip509 = match Cip509::decode(&mut decoder, &mut report) {
+            Ok(v) => v,
+            Err(e) => {
```
Agree, if there is just no CIP509 data, we should error, otherwise the caller has no way to know there was no data present vs there was data but it had problems.

stanislav-tkach replied:
Fixed.

stevenj commented on rust/rbac-registration/src/cardano/cip509/rbac/metadata.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+/// Decodes an array of type T.
+fn decode_array<'b, T>(d: &mut Decoder<'b>, context: &str, report: &mut ProblemReport) -> Vec<T>
```
At a low level point like this, if the array can't decode you no longer have any state to allow decoding to continue.
I do think this should update the problem report, and then error, so that the decoder can realize its done as much as it can do.

stanislav-tkach replied:
I agree with this, but I think that mixing different error handling strategies is hard to manage. 

For example: we were able to successfully decode `purpose` and `txn_inputs_hash`, but unable to decode `prv_tx_id`. It makes sense to stop the decoding because we will be unable to read any other fields anyway, but it is very easy to (mistakenly?) propagate `decode::Error` all the way to the top. In this case we discard the problem report that can still contain valuable messages such as purpose having an invalid length (valid CBOR array of bytes, but too few or too many bytes for `UUID`). On the other hand, we can currently generate multiple report entries for the same failure. For example, "unable to decode some field", "missing required field", etc. But I still think it is better than losing the information.

Also we perform the validation after the decoding. This is not ideal, either, but it isn't always possible to do it right after decoding the specific data because validation can depend on other fields.

In conclusion I'm also not fully happy with the current state of the error handling, but I don't see how this can be significantly improved. I will think more about this, but I would prefer not to drag this pull request endlessly and maybe make some incremental improvements separately.

stanislav-tkach replied:
I think I fixed that. 🙃

stevenj reviewed (changes requested):
Small suggestions from me.

stanislav-tkach commented:
Despite my previous comment, I think I was able to improve the error handling, so now we stop decoding when a "fatal" error is encountered, but still preserve the decoded data.

stanislav-tkach commented:
I think this is (once again) is ready for the (final?) review.

stanislav-tkach commented:
The tests data needs to be updates again for tests to pass.

bkioshn commented on rust/rbac-registration/src/cardano/cip509/validation.rs:
```diff
@@ -21,659 +21,348 @@
+
+/// Checks that only role 0 uses certificates with zero index.
+#[allow(clippy::similar_names)]
+pub fn validate_role_data(metadata: &Cip509RbacMetadata, report: &ProblemReport) {
```
For None role 0, 
if there is Some x509 or is Some c509 or is Some pubkey , the first index should be set to undefined

Should not able to do delete in index 0 since it is consider an update to role 0

bkioshn commented on rust/rbac-registration/src/cardano/cip509/validation.rs:
```diff
@@ -21,659 +21,348 @@
 
-use c509_certificate::{general_names::general_name::GeneralNameValue, C509ExtensionType};
-use der_parser::der::parse_der_sequence;
+use std::borrow::Cow;
```
Maybe we can delete the ^ docs and add it into the CIP document instead

bkioshn commented on rust/rbac-registration/src/cardano/transaction/raw_aux_data.rs:
```diff
@@ -41,19 +40,19 @@ impl RawAuxData {
 
         match decoder.datatype() {
-            Ok(minicbor::data::Type::Map) => {
+            Ok(Type::Map) => {
```
This file can be safely remove, since the logic is moved to `cardano-blockchain-types`

bkioshn commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,581 @@
+            return Ok(None);
+        };
+
+        let auxiliary_data = match &txn.auxiliary_data {
+            Nullable::Some(v) => v.raw_cbor(),
+            _ => return Ok(None),
+        };
+        let raw_auxiliary_data = RawAuxData::new(auxiliary_data);
+        let Some(metadata) = raw_auxiliary_data.get_metadata(LABEL) else {
+            return Ok(None);
+        };
```
Can now do
```
 let metadata = block.txn_metadata(index, MetadatumLabel::CIP509_RBAC);
 let aux = block.aux_data().get(index);
 ```
 Might need to add a function to get the raw aux

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,572 @@
+    /// A registration purpose (`UUIDv4`).
+    ///
+    /// The purpose is defined by the consuming dApp.
+    purpose: Option<Uuid>,
```
I think it could you use a `UuidV4` type from the `catalyst-types` crate, so we will be totally sure that its a uuid v4.

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/cip509.rs:
```diff
@@ -0,0 +1,572 @@
+    ///
+    /// The data located in `Cip509` is only considered valid if
+    /// `ProblemReport::is_problematic()` returns false.
+    report: ProblemReport,
```
Seems redundant at this place as we are storing it inside the `DecodeContext` struct.
Also by storing problem report at this place we assuming that it could be a `Cip509` invalidly structured object, what I think its better to avoid and keep with the following logic.
- trying to decode bytes to `Cip509`.
- if it fails getting an error and problem report, not getting a `Cip509` object.
- if it is ok, getting a valid (at least structurally) `Cip509`.

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/validation.rs:
```diff
@@ -21,659 +21,348 @@
-    None
-}
+/// Checks that the role 0 data is correct.
+fn validate_role_0(
```
Clarified: there should be no encryption key for the role 0.

stanislav-tkach commented on rust/rbac-registration/Cargo.toml:
```diff
@@ -26,11 +26,12 @@ brotli = "7.0.0"
-c509-certificate = { version = "0.0.3", git = "https://github.com/input-output-hk/catalyst-libs.git" , tag = "v0.0.3" }
+c509-certificate = { version = "0.0.3", git = "https://github.com/input-output-hk/catalyst-libs.git", tag = "v0.0.3" }
 pallas = { version = "0.30.1", git = "https://github.com/input-output-hk/catalyst-pallas.git", rev = "9b5183c8b90b90fe2cc319d986e933e9518957b3" }
+cardano-blockchain-types = { version = "0.0.1", path = "../cardano-blockchain-types" }
```
As far as I can see, we use tagged version everywhere else, so it is my oversight.

stanislav-tkach replied:
I will fix that after https://github.com/input-output-hk/catalyst-libs/pull/172 is merged.

stanislav-tkach replied:
I used a specific revision for the `cbork-utils` crate, but it is currently impossible to use a revision for `cardano-blockchain-types` and `catalyst-types` because this pull request changes them. I can move these changes to a separate pull request, but I'm not sure if it is worth doing.

bkioshn commented on rust/rbac-registration/src/cardano/cip509/validation.rs:
```diff
@@ -1,679 +1,409 @@
+        None | Some(SimplePublicKeyType::Undefined)
+    ) {
+        report.other(
+            "The public key cannot be used for the role 0, only a certificate",
```
Since this is checking for other role too. Maybe better to say
"Role 0 requires a certificate, not a public key; for other roles, set the first public key to undefined if needed."

Or add this check for role 0 and other role so the error messages are distinct

Mr-Leshiy reviewed (approved):
LGTM
